Houdini symbol visibility management

andrewkaufman opened the pull request:
This caries on from #695 applying symbol visibility to IECoreHoudini.

Along the way I (re)discovered that libIECoreHoudini has a python dependency, which isn't ideal, so I removed one of those and added todos for the rest.

I was hoping this would help me along on the issues I'm hitting in #954, but alas it did not. There is enough here that it seemed worth splitting out to its own PR, as the longer 07d552e lingers the harder it will be to merge.

Tests are passing in H17.0.416, @lucienfostier is going to verify in H17.5 as well.

Note: I don't really understand symbol visibility too deeply. I'm more or less following the pattern established in #695... but I have verified there are fewer public symbols in the resulting libIECoreHoudini.so and _IECoreHoudini.so (bindings), and I did hit a couple missing symbols that were needed along the way, so I'm fairly confident its working.

andrewkaufman commented:
@lucienfostier confirms tests also pass with H17.5.258

johnhaddon commented on src/IECoreHoudini/SOP_OpHolder.cpp:
```diff
@@ -143,6 +143,8 @@ void SOP_OpHolder::doOperation( IECore::Op *op, const GU_DetailHandle &handle, c
 	}
 	catch( boost::python::error_already_set )
 	{
+		// \todo: libIECoreHoudini should not use python. Determine if
```
It looks like the real issue here is that `IECorePython::OpWrapper::doOperation` isn't doing any exception management. It should be catching `error_already_set` itself and then using `ExceptionAlgo::translatePythonException()` to convert it to a regular exception that would be covered by the cases below. This is worth doing whether or not IECoreHoudini retains a dependency on python.

andrewkaufman replied:
Thanks for finding that. I spawned an internal ticket for it.

johnhaddon commented on SConstruct:
```diff
@@ -2471,6 +2471,8 @@ houdiniEnvAppends = {
 		"-std=$CXXSTD",
 	],
 	"CPPFLAGS" : [
+		## \todo: libIECoreHoudini should not use python.
```
Quick question : does Houdini itself have a core dependency on Python? If so, what's the harm in IECoreHoudini having one?

andrewkaufman replied:
I believe you can write Houdini executables that don't use python. I'm not sure anyone actually does that, but I think the same could be said of Gaffer, right? To me, its more about following our conventions for code separation. In the context of this PR, I wish I was able to be sure that the "missing to_python by value converter" I'm trying to chase down was not caused by libIECoreHoudini.

johnhaddon replied:
You can make an executable linking against libGaffer without needing to pull in libpython, yes - the separation is more than just convention. A quick look at Houdini's libs suggests the same isn't the case there though, so I was just wondering if trying to follow convention for IECoreHoudini actually had similar benefits. Not having to worry about `boost::python` at all sounds worth the effort though...

andrewkaufman replied:
I'm surprised to see you're right, 19 of the 21 Houdini libs link to python (and 6 of those to hboost_python). I thought they used to be better about the separation as well, with the aim of having standalone executables using Houdini libs...

Anyways, as you agree, its worth removing the dependency to avoid `boost::python` issues

johnhaddon commented on src/IECoreHoudini/bindings/FnParameterisedHolderBinding.cpp:
```diff
@@ -36,13 +36,154 @@
+namespace
+{
+
+class FnParameterisedHolder
```
Is there no possibility of wanting to use an FnParameterisedHolder from C++?

andrewkaufman replied:
The comment on the original header was `/// \todo: This class is a bit funny. Can this all be moved to the bindings like in IECoreMaya?`.

I looked at the Maya equivalents and the only `IECoreMaya` namespaced functions are `parameterisedHolderAssignModificationState` and `bindFnParameterisedHolder`. There are a bunch of functions that aren't anonymous namspaced, but I'd assumed that's because they were written before we adopted that convention.

Also, this is just wrapping functions from `IECoreHoudini::ParameterisedHolder` with an "alive" flag, so a C++ user can use the node directly without needing the `Fn` and ask if its alive using the HDK (c++ api).

johnhaddon replied:
> this is just wrapping functions from IECoreHoudini::ParameterisedHolder with an "alive" flag, so a C++ user can use the node directly without needing the Fn and ask if its alive using the HDK (c++ api).

Ah - got it. I'd looked at SOP_ParameterisedHolder and noted it didn't have `setParameterised()` but then hadn't taken the time to realise it got it from the base class. Thanks for clarifying.

johnhaddon commented on include/IECoreHoudini/CoreHoudini.h:
```diff
@@ -57,7 +59,7 @@ class CoreHoudini
 
 		/// This loads hou into the global context
-		static void initPython();
+		static void IECOREHOUDINI_API initPython();
```
Any reason not to just mark the whole class with IECOREHOUDINI_API?

andrewkaufman replied:
Just to limit the number of symbols I'm exposing, with a mind of getting rid of this entire thing eventually. I could change it if you prefer though.

johnhaddon replied:
I think logically the class is part of the API, so it would make more sense to expose the whole thing, even if we want to remove it later.

andrewkaufman replied:
Fixed in 26177e044b44ce9bbcdde02405220876d66db3b4

johnhaddon commented on SConstruct:
```diff
@@ -2461,14 +2461,14 @@ if doConfigure :
 
 houdiniEnvSets = {
 	"IECORE_NAME" : "IECoreHoudini",
-	"CXXFLAGS" : dependencyIncludes + [ "$HOUDINI_CXX_FLAGS", "-DMAKING_DSO" ]
```
Just checking I understand : it's this change from set to append that means the `-fvisibility=hidden` compiler flag gets set when building IECoreHoudini, correct?

andrewkaufman replied:
Yes, that's correct. As well as the other flags we were setting above like `-Werror`, the BUILD_TYPE flags, extra ASAN flags, etc.

andrewkaufman commented:
I've squashed in the `IECOREHOUDINI_API` change, rebased on master to incorporate Houdini changes from #958, and re-run testHoudini for 17.0.416 and 17.5.258

johnhaddon commented:
LGTM
